Rollback a primary before starting to recover from translog

dnhatn opened the pull request:
Today we always recover a primary from the last commit point. However with a new deletion policy, we keep multiple commit points in the existing store, thus we have chance to find a good starting commit point. With a good starting commit point, we may be able to throw away stale operations. This PR rollbacks a primary to a starting commit then recovering from translog.

Relates #10708

dnhatn commented:
> I don't think we should delete commits.

@bleskes I agree. I've updated the PR to let an engine open a starting commit point. Could you please take a look? Thank you.

bleskes commented on core/src/main/java/org/elasticsearch/index/shard/IndexShard.java:
```diff
@@ -2180,7 +2184,15 @@ private EngineConfig newEngineConfig(EngineConfig.OpenMode openMode) {
+                throw new AssertionError("unknown recovery type: [" + recoveryType + "]");
+        }
+        final IndexCommit startingCommit;
+        if (recoveryType == RecoverySource.Type.EXISTING_STORE) {
```
why does this needs to be out of engine? can't we do it in the engine construct as an invariant when opening an index (and translog)

dnhatn replied:
Yep, I will it to the Engine.

bleskes commented on core/src/main/java/org/elasticsearch/index/engine/CombinedDeletionPolicy.java:
```diff
@@ -90,12 +101,38 @@ private void updateTranslogDeletionPolicy(final IndexCommit minRequiredCommit, f
+        // Snapshotted commits may not have all its required translog.
+        final List<IndexCommit> recoverableCommits = new ArrayList<>();
+        for (IndexCommit commit : commits) {
+            if (minRetainedTranslogGen <= Long.parseLong(commit.getUserData().get(Translog.TRANSLOG_GENERATION_KEY))) {
```
why do we do this? isn't the logic in indexOfKeptCommits enough to deal with this?

dnhatn replied:
In the previous 6.x versions, we keep the last commit and translog for that commit only. If we take a snapshot and commit, we will have two commits but translog for the last commit only. During the store recovery, if the max_seqno of the last commit is greater than the global checkpoint, the Policy will pick the snapshotted commit although it does not have full translog.

bleskes commented on core/src/main/java/org/elasticsearch/index/engine/CombinedDeletionPolicy.java:
```diff
@@ -62,7 +66,14 @@ public void onInit(List<? extends IndexCommit> commits) throws IOException {
+                } else {
+                    assert commits.contains(startingIndexCommit) : "Existing commits must contain the starting commit; " +
+                        "startingCommit [" + startingIndexCommit + "], commits [" + commits + "]";
+                    commits.stream().filter(commit -> startingIndexCommit.equals(commit) == false).forEach(IndexCommit::delete);
```
why do we need special handling here and need the start commit point? can you explain?

dnhatn replied:
I calculated the retained translog generations incorrectly; I will revert this change. There is an issue with the local checkpoint; I will reach out to discuss with you @bleskes.

dnhatn commented:
@bleskes, I've moved the starting commit point to the InternalEngine and removed the special case in the CombinedDeletionPolicy. Would you please take another look? Thank you.

bleskes commented on core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java:
```diff
@@ -237,23 +243,24 @@ public InternalEngine(EngineConfig engineConfig) {
+            default:
+                throw new IllegalArgumentException("unknown type: " + openMode);
         }
-        return localCheckpointTrackerSupplier.apply(maxSeqNo, localCheckpoint);
```
i like the final approach better this gives a uniform return value (easier to understand) and it makes sure these things are set.

dnhatn commented on core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java:
```diff
@@ -179,6 +182,9 @@ public InternalEngine(EngineConfig engineConfig) {
             mergeScheduler = scheduler = new EngineMergeScheduler(engineConfig.getShardId(), engineConfig.getIndexSettings());
             throttle = new IndexThrottle();
             try {
+                this.startingCommit = getStartingCommitPoint();
```
It is used in two other places: #recoverFromTranslogInternal and #createLocalCheckpointTracker.

bleskes replied:
I think it should be a parameter to createLocalCheckpointTracker and I left a comment on recoverFromTranslogInternal . The less state the better ;)

bleskes commented on core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java:
```diff
@@ -401,10 +411,21 @@ public InternalEngine recoverFromTranslog() throws IOException {
         Translog.TranslogGeneration translogGeneration = translog.getGeneration();
         final int opsRecovered;
-        final long translogGen = Long.parseLong(lastCommittedSegmentInfos.getUserData().get(Translog.TRANSLOG_GENERATION_KEY));
+        final long translogGen = Long.parseLong(startingCommit.getUserData().get(Translog.TRANSLOG_GENERATION_KEY));
```
I think the lastCommittedSegmentInfo should be set to the opening commit, no?

dnhatn commented:
@bleskes, I have addressed your feedbacks. Would you please give it another go. Thank you.

dnhatn commented:
please test this.

bleskes commented on core/src/main/java/org/elasticsearch/index/engine/CombinedDeletionPolicy.java:
```diff
@@ -90,12 +91,26 @@ private void updateTranslogDeletionPolicy(final IndexCommit minRequiredCommit, f
 
+    /**
+     * Find a safe commit point from a list of existing commits based on the persisted global checkpoint from translog.
+     * The max seqno of a safe commit point should be at most the global checkpoint from the translog checkpoint.
```
can we add a huge warning that says that if the index was created before 6.2 we can't guarantee we'll find one and that in that case we return the oldest valid commit?

bleskes commented on core/src/main/java/org/elasticsearch/index/engine/CombinedDeletionPolicy.java:
```diff
@@ -90,12 +91,26 @@ private void updateTranslogDeletionPolicy(final IndexCommit minRequiredCommit, f
     }
 
+    /**
+     * Find a safe commit point from a list of existing commits based on the persisted global checkpoint from translog.
```
persisted global checkpoint from the translog -> supplied global checkpoint.

bleskes commented on core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java:
```diff
@@ -177,14 +180,17 @@ public InternalEngine(EngineConfig engineConfig) {
             try {
-                this.localCheckpointTracker = createLocalCheckpointTracker(localCheckpointTrackerSupplier);
+                final IndexCommit startingCommit = getStartingCommitPoint();
+                assert startingCommit == null || openMode == EngineConfig.OpenMode.OPEN_INDEX_AND_TRANSLOG :
```
I don't think this boolean does what you want? if start commit is null it will always pass.

bleskes commented on core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java:
```diff
@@ -243,8 +249,14 @@ private LocalCheckpointTracker createLocalCheckpointTracker(
                 break;
             case OPEN_INDEX_AND_TRANSLOG:
+                // When recovering from a previous commit point, we use the local checkpoint from that commit,
+                // but the max_seqno from the last commit. This allows use to throw away stale operations.
```
why do we need this distinction? can't we keep it conceptually simple? also we recover everything from the translog, so I'm not sure how it can happen that we throw away stuff?

dnhatn replied:
- If we use the local checkpoint from the last commit, operations having seqno less than or equal to the local checkpoint will be skipped -> we need to use the local checkpoint from the starting commit.

- We fillSeqNoGaps until the max_seqno -> we need to use max_seqno from the last commit to have full history.

bleskes replied:
> We fillSeqNoGaps until the max_seqno -> we need to use max_seqno from the last commit to have full history.

We do so after we replay the translog. at which point we'll know what the max seq is (and it may be higher than the one in the last commit, but it can't be lower)

bleskes commented on core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java:
```diff
@@ -399,6 +411,32 @@ public InternalEngine recoverFromTranslog() throws IOException {
+                        recoverableCommits.add(commit);
+                    }
+                }
+                assert recoverableCommits.isEmpty() == false : "Unable to select a proper safe commit point; " +
```
I think you want to say that no commit point was found which could be recovered from the translog.

bleskes commented on core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java:
```diff
@@ -399,6 +411,32 @@ public InternalEngine recoverFromTranslog() throws IOException {
+            // To avoid this issue, we only select index commits whose translog files are fully retained.
+            if (engineConfig.getIndexSettings().getIndexVersionCreated().before(Version.V_6_2_0)) {
+                final List<IndexCommit> recoverableCommits = new ArrayList<>();
+                final long minRetainedTranslogGen = Translog.readMinReferencedTranslogGen(translogPath);
```
since we open the translog before the writer now - can we use the open translog for all this information rather than static reading it? this will make sure that we use something that went through all the right validations.

bleskes commented on core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java:
```diff
@@ -521,7 +559,11 @@ private ExternalSearcherManager createSearcherManager(SearchFactory externalSear
                 internalSearcherManager = new SearcherManager(directoryReader,
                         new RamAccountingSearcherFactory(engineConfig.getCircuitBreakerService()));
-                lastCommittedSegmentInfos = readLastCommittedSegmentInfos(internalSearcherManager, store);
+                if (openMode == EngineConfig.OpenMode.OPEN_INDEX_AND_TRANSLOG) {
```
can you explain why we need this destinction?

dnhatn replied:
I think if we open other commit rather the last commit, we should assign lastCommittedSegmentInfos from that commit. I can revert this change and add this logic to `recoverFromTranslogInternal`. Your thought?

bleskes commented on core/src/test/java/org/elasticsearch/index/engine/InternalEngineTests.java:
```diff
@@ -1121,8 +1120,12 @@ public void testRenewSyncFlush() throws Exception {
     }
 
     public void testSyncedFlushSurvivesEngineRestart() throws IOException {
+        final LongSupplier inSyncGlobalCheckpointSupplier = () -> this.engine.getLocalCheckpointTracker().getCheckpoint();
```
why is this needed?

bleskes commented on core/src/test/java/org/elasticsearch/index/shard/IndexShardTests.java:
```diff
@@ -1092,6 +1086,7 @@ public void testAcquireIndexCommit() throws Exception {
     public void testSnapshotStore() throws IOException {
         final IndexShard shard = newStartedShard(true);
         indexDoc(shard, "test", "0");
+        shard.updateLocalCheckpointForShard(shard.shardRouting.allocationId().getId(), 0);
```
I presume you did this to make sure the global checkpoint advances. I wonder if we should fold into indexDoc? it's sneaky

bleskes commented on core/src/test/java/org/elasticsearch/index/shard/IndexShardTests.java:
```diff
@@ -1745,6 +1758,37 @@ public void testFailIfIndexNotPresentInRecoverFromStore() throws Exception {
+        flushShard(shard);
+        assertThat(getShardDocUIDs(shard), containsInAnyOrder("doc-0", "doc-1"));
+        // Simulate resync (without rollback): Noop #1, index #2
+        shard.markSeqNoAsNoop(1, "test");
```
I *think* you need a primary term of 2 here?

bleskes commented on core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java:
```diff
@@ -521,7 +550,7 @@ private ExternalSearcherManager createSearcherManager(SearchFactory externalSear
                 internalSearcherManager = new SearcherManager(directoryReader,
                         new RamAccountingSearcherFactory(engineConfig.getCircuitBreakerService()));
-                lastCommittedSegmentInfos = readLastCommittedSegmentInfos(internalSearcherManager, store);
+                lastCommittedSegmentInfos = store.readCommittedSegmentsInfo(indexWriter.getConfig().getIndexCommit());
```
add a comment that `indexWriter.getConfig().getIndexCommit()` can be null?

bleskes commented on core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java:
```diff
@@ -1778,9 +1807,11 @@ private long loadCurrentVersionFromIndex(Term uid) throws IOException {
         try {
             final IndexWriterConfig iwc = getIndexWriterConfig(create);
+            assert startingCommit == null || create == false : "Starting commit makes sense only when create=false";
+            iwc.setIndexCommit(startingCommit);
```
this should be part of getIndexWriterConfig?

bleskes commented on core/src/test/java/org/elasticsearch/index/engine/InternalEngineTests.java:
```diff
@@ -2514,6 +2518,7 @@ private Mapping dynamicUpdate() {
     }
 
     public void testTranslogReplay() throws IOException {
+        final LongSupplier inSyncGlobalCheckpointSupplier = () -> this.engine.getLocalCheckpointTracker().getCheckpoint();
```
I'm wondering if we should make this the default?

dnhatn replied:
Most of the time, we prefer to deal with an in-sync global checkpoint, thus making this as a default makes sense to me. I will do in a follow-up.

dnhatn commented:
@bleskes, Thanks a lot for your review.
